flink-runtime/src/main/java/org/apache/flink/runtime/client/JobClient.java

```diff
@@ -118,27 +138,162 @@ public static JobExecutionResult submitJobAndWait(
 			sysoutLogUpdates);
 
 		ActorRef jobClientActor = actorSystem.actorOf(jobClientActorProps);
-		
+
+		Future<Object> submissionFuture = Patterns.ask(
+				jobClientActor,
+				new JobClientMessages.SubmitJobAndWait(jobGraph),
+				new Timeout(AkkaUtils.INF_TIMEOUT()));
+
+		return new JobListeningContext(
+				jobGraph.getJobID(),
+				submissionFuture,
+				jobClientActor,
+				classLoader);
+	}
+
+
+	/**
+	 * Attaches to a running Job using the JobID.
+	 * Reconstructs the user class loader by downloading the jars from the JobManager.
+	 * @throws JobRetrievalException if anything goes wrong while retrieving the job
+	 */
+	public static JobListeningContext attachToRunningJob(
+			JobID jobID,
+			ActorGateway jobManagerGateWay,
+			Configuration configuration,
+			ActorSystem actorSystem,
+			LeaderRetrievalService leaderRetrievalService,
+			FiniteDuration timeout,
+			boolean sysoutLogUpdates) throws JobRetrievalException {
+
+		checkNotNull(jobID, "The jobID must not be null.");
+		checkNotNull(jobManagerGateWay, "The jobManagerGateWay must not be null.");
+		checkNotNull(configuration, "The configuration must not be null.");
+		checkNotNull(actorSystem, "The actorSystem must not be null.");
+		checkNotNull(leaderRetrievalService, "The jobManagerGateway must not be null.");
+		checkNotNull(timeout, "The timeout must not be null.");
+
+		// retrieve classloader first before doing anything
+		ClassLoader classloader;
+		try {
+			classloader = retrieveClassLoader(jobID, jobManagerGateWay, configuration, timeout);
```

Comment:
In addition to the result, you'll also need the class loader for getting accumulators of a running job. 

I agree that it would be nice to fail when the class loader can't be reconstructed, but *only* if it is really the only option. So we could start off with the class loader set to `None` in the `JobListeningContext`. When the class loader is needed, i.e. accumulator retrieval or job execution result retrieval, it is fetched.